src/spotify-service.ts

```diff
@@ -1,67 +1,78 @@
-    Effect.gen(function* () {
-      yield* Effect.logInfo("starting spotify service");
-      const api = yield* SpotifyApiClient;
-      const pubsub = yield* MessagePubSub;
-
-      const dequeue = yield* pubsub.subscribeTo("CurrentlyPlayingRequest");
+export const SpotifyService = Layer.scopedDiscard(
+  Effect.gen(function* () {
+    yield* Effect.logInfo("starting spotify service");
+    const api = yield* SpotifyApiClient;
+    const pubsub = yield* MessagePubSub;
 
-      yield* Effect.forkScoped(
-        Effect.forever(
-          Effect.gen(function* () {
-            yield* Effect.logInfo("starting CurrentlyPlayingRequest listener");
-            // todo: reccommend takeWhen
-            yield* Queue.take(dequeue);
-            yield* Effect.logInfo("received CurrentlyPlayingRequest listener");
+    const dequeue = yield* pubsub.subscribeTo("CurrentlyPlayingRequest");
 
-            const { item } = yield* Effect.tryPromise(() =>
-              api.player.getCurrentlyPlayingTrack(undefined),
-            );
+    yield* Effect.forkScoped(
+      Effect.forever(
+        Effect.gen(function* () {
+          yield* Effect.logInfo("starting CurrentlyPlayingRequest listener");
+          // todo: reccommend takeWhen
+          yield* Queue.take(dequeue);
+          yield* Effect.logInfo("received CurrentlyPlayingRequest listener");
 
-            yield* Effect.logInfo("resolved spotify api request");
+          const { item } = yield* api.use((_) =>
+            _.player.getCurrentlyPlayingTrack(undefined)
+          );
 
-            if (!("album" in item)) {
-              yield* Effect.logWarning(`Invalid Spotify Track Item`);
-              return yield* Effect.void;
-            }
+          yield* Effect.logInfo("resolved spotify api request");
 
-            yield* Effect.logInfo("publishing currently playing");
-            yield* pubsub.publish(Message.CurrentlyPlaying({ song: item }));
-
-            return yield* Effect.void;
-          }),
-        ),
-      );
+          if (!("album" in item)) {
+            yield* Effect.logWarning(`Invalid Spotify Track Item`);
+            return;
+          }
 
-      return yield* Effect.never;
-    }),
-  ).pipe(Effect.provide(Layer.mergeAll(SpotifyApiClient.Live))),
-);
+          yield* Effect.logInfo("publishing currently playing");
+          yield* pubsub.publish(Message.CurrentlyPlaying({ song: item }));
+        })
+      )
+    );
+  })
+).pipe(Layer.provide(SpotifyApiClient.Live), Layer.provide(MessagePubSub.Live));
 
 export function requestAccessToken(code: string) {
   return Effect.gen(function* () {
     const config = yield* SpotifyConfigService;
-    const authorizationHeader = `Basic ${Encoding.encodeBase64(`${config.clientId}:${Secret.value(config.clientSecret)}`)}`;
+    const authorizationHeader = `Basic ${Encoding.encodeBase64(
+      `${config.clientId}:${Secret.value(config.clientSecret)}`
+    )}`;
 
     // TODO: Refactor to platform HttpClient
```

Comment:
https://github.com/tim-smart/effect-twitch-integrations/blob/platform-htp/src/spotify-service.ts#L83

Reply:
This is slick 🔥